python/tutorials/03-matrix-multiplication-all-types.py

```diff
@@ -0,0 +1,532 @@
+        (d_type == tl.float8e5b16 and TORCH_HAS_FP8E5B16) or not d_type.is_fp8():
+        input = raw_data.to(tl_to_torch_types[d_type])
+        input_f16 = input.to(torch.float16)
+    else:
+        f8_tensor = raw_data.to(torch.int8)
+        # keep only two bits of exponent to avoid overflow
+        f8_tensor = f8_tensor & 0b00111111
+        input = triton.reinterpret(f8_tensor, d_type)
+        input_f16 = torch.empty_like(f8_tensor, dtype=torch.float16)
+        grid = lambda meta: (triton.cdiv(n_elements, meta['BLOCK_SIZE']),)
+        n_elements = raw_data.numel()
+        copy_kernel[grid](input, input_f16, n_elements, BLOCK_SIZE=1024)
+
+    return input, input_f16
+
+# %%
+# Unit Test
+# ---------
+#
+# We can test our custom matrix multiplication operation against a native torch implementation (i.e., cuBLAS).
+@pytest.mark.parametrize("M, N, K, in_dtype, out_dtype",
+[ (*shape, in_dtype, out_dtype)
+    for shape in [(128, 256, 32), (128, 16, 32), (32, 128, 64),
+                  (128, 128, 64), (64, 128, 128), (32, 128, 64),
+                  (64, 64, 32), (32, 32, 128), (128, 128, 64),
+                   (64, 128, 128), (512, 512, 512), (1024, 1024, 1024)]
+    for in_dtype, out_dtype in [('fp16', 'fp16'),
+                                ('bf16', 'bf16'),
+                                ('fp16', 'fp32'),
+                                ('fp32', 'fp32'),
+                                ('fp8e4', 'fp16'),
+                                ('fp8e5', 'fp16'),
+                                ('int8', 'int8'),
+                                ('int8', 'int32')]]
+)
+def test_correctness(M, N, K, in_dtype, out_dtype):
+    a, a_fp16 = gen_input(M, K, in_dtype, 1, device='cuda')
+    b, b_fp16 = gen_input(K, N, in_dtype, 2, device='cuda')
+    # Allocates output.
+    tl_out_dtype = name_to_tl_types[out_dtype]
+    c = torch.empty((M, N), device=a.device, dtype=tl_to_torch_types[tl_out_dtype])
+    matmul(a, b, c, activation="")
+    torch_output = torch.matmul(a_fp16, b_fp16)
+    print(f"triton_output={c}")
+    print(f"torch_output={torch_output}")
+    rtol = 0 if torch.version.hip is None else 1e-2
+    if torch.allclose(c.to(torch.float16), torch_output, atol=5e-2, rtol=rtol):
+        print("✅ Triton and Torch match")
+    else:
+        print("❌ Triton and Torch differ")
+        assert torch.allclose(c, torch_output, atol=1e-2, rtol=rtol)
+
+
+# %%
+# Benchmark
+# ---------
+#
+# Square Matrix Performance
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~
+#
+# We can now compare the performance of our kernel against that of cuBLAS. Here we focus on square matrices,
+# but feel free to arrange this script as you wish to benchmark any other matrix shape.
+
+def get_type(provider):
+    res = re.findall(r'\(.*?\)', provider)
+    return res[0][1:-1]
+
+def get_x_vals():
+    x_vals = [(512 * v, 512 * v, 512 * v) for v in range (2, 17)]
+
+    x_vals += [(4864, 4096, 8192), (9728, 8192, 65536)]
```

Comment:
Question: This isn't square. Why this size? Is this the "optimal MI300" size?

Reply:
I think 4864, 4096, 8192 is the optimal MI300 size.